plugin/plugin.go

```diff
@@ -0,0 +1,248 @@
+		}
+	}
+	return nil
+}
+
+func (p *FlywayPlugin) Run() error {
+	var stdoutBuf, stderrBuf bytes.Buffer
+	var err error
+
+	p.ExecCommand = p.GetExecArgsStr()
+	logrus.Infof("Executing command: %s", p.ExecCommand)
+
+	cmdParts := strings.Fields(p.ExecCommand)
+	if len(cmdParts) < 2 {
+		return fmt.Errorf("Invalid command: %s", p.ExecCommand)
+	}
+	cmdName := cmdParts[0]
+	cmdArgs := cmdParts[1:]
+
+	cmd := exec.Command(cmdName, cmdArgs...)
+	cmd.Stdout = &stdoutBuf
+	cmd.Stderr = &stderrBuf
+	if len(p.InputArgs.DriverPath) > 0 {
+		p.Env = "CLASSPATH=" + p.InputArgs.DriverPath
+		cmd.Env = append(os.Environ(), "CLASSPATH="+p.InputArgs.DriverPath)
+	}
+
+	if p.InputArgs.IsDryRun == "TRUE" {
+		return nil
+	}
+
+	err = cmd.Run()
+	if err == nil {
+		logrus.Println(stdoutBuf.String())
+		logrus.Infof("Command executed successfully: %s", stdoutBuf.String())
+	} else {
+		logrus.Println(stderrBuf.String())
+		logrus.Errorf("Command execution failed: %s, error: %v", stderrBuf.String(), err)
+	}
+
+	return nil
+}
+
+func (p *FlywayPlugin) GetExecArgsStr() string {
+	var builder strings.Builder
+
+	builder.WriteString(GetFlywayExecutablePath() + " ")
+	builder.WriteString(p.InputArgs.FlywayCommand + " ")
+	builder.WriteString(p.CommandSpecificArgs + " ")
+
+	if len(p.InputArgs.Url) > 0 {
+		builder.WriteString("-url=" + p.InputArgs.Url + " ")
+	}
+	if len(p.InputArgs.UserName) > 0 {
+		builder.WriteString("-user=" + p.InputArgs.UserName + " ")
+	}
+	if len(p.InputArgs.Password) > 0 {
+		builder.WriteString("-password=" + p.InputArgs.Password + " ")
+	}
+	if len(p.InputArgs.Locations) > 0 {
+		builder.WriteString("-locations=" + p.InputArgs.Locations + " ")
+	}
+	// this should be the last
+	builder.WriteString(p.InputArgs.CommandLineArgs)
+
+	return builder.String()
+}
+
+func (p *FlywayPlugin) IsCommandValid() error {
+	if p.InputArgs.FlywayCommand == "" {
+		return fmt.Errorf("Command is empty")
+	}
+
+	err := p.IsUnknownCommand()
+	if err != nil {
+		return err
+	}
+
+	err = p.CheckMandatoryArgs()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *FlywayPlugin) IsUnknownCommand() error {
+	if _, ok := knownCommandsMap[p.InputArgs.FlywayCommand]; !ok {
+		return fmt.Errorf("unknown command: %s", p.InputArgs.FlywayCommand)
+	}
+	return nil
+}
+
+func (p *FlywayPlugin) CheckMandatoryArgs() error {
+
+	args := p.InputArgs
+
+	if strings.Contains(args.CommandLineArgs, ConfigFileOpt) { // pick args from file
+		return nil
+	}
+
+	type mandatoryArg struct {
+		EnvName   string
+		ParamName *string
+		Hint      string
+	}
+
+	ma := []mandatoryArg{
+		{"FLYWAY_URL", &args.Url, "url"},
+		{"FLYWAY_USER", &args.UserName, "username"},
+		{"FLYWAY_PASSWORD", &args.Password, "password"},
+	}
+
+	for _, m := range ma {
+		if os.Getenv(m.EnvName) == "" && *m.ParamName == "" {
+			LogPrintln("Missing mandatory argument: " + m.EnvName)
+			return fmt.Errorf("Missing mandatory argument: %s (env: %s)", m.Hint, m.EnvName)
+		}
+	}
```

Comment:
```suggestion
	for _, m := range ma {
	        value := os.Getenv(m.EnvName)
		if value == "" && *m.ParamName == "" {
			eturn fmt.Errorf("missing mandatory argument: %s", m.Hint)
		}
		*m.ParamName = value
	}
```

Reply:
for _, m := range ma {
		if os.Getenv(m.EnvName) == "" && m.ParamName == "" {
			LogPrintln("Missing mandatory argument: " + m.EnvName)
			return fmt.Errorf("Missing mandatory argument: %s (env: %s)", m.Hint, m.EnvName)
		}
	} is just local check not adding *m.ParamName = value. Also struct mandatoryArg pointer type *ParamName to ParamName. Pointer may seem like that value is to be modified, but read only pass by value is sufficient